accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -410,24 +417,24 @@ read.
     is treated as a first read and the initializer expression is evaluated
     again.
 
-A write to a field or variable which is marked `final` and `late` is a runtime
+A write to a field or variable which is marked `late` and `final` is a runtime
```

Comment:
Is this a *dynamic* write?

Otherwise, if the variable has an initializer expression, then it's a *compile-time* error, not a run-time error.

SO maybe:

> A write to a field or variable which is declared `late` and `final` and with no intiializer expression, is a run-time error if there has been a previous write to the field or variable (including ...).

Reply:
It is a compile time error if it is statically visible as such, but you can arrange (via overriding) for it not to be visible as such.  I'm not sure I follow the gist of your proposed change?

Reply:
I agree that my change technically says the same thing. At least after reading things a few extra times.

I guess my problem is, in general, with rules stated as "something happens, unless (long list of exceptions covering most actual situations)". I just find it harder to grok. Especially when some of those exceptions are compile-time errors, where it feels wrong to talk about run-time semantics of the code at all.

I am still not entirely convinced that the current text is not subtly wrong.
It seems to say that it is not a run-time error to write to a late final field which has an initializer expression. 
But you *cannot* write to such a field, it has no setter, so saying what happens or doesn't happen *when you write* is meaningless. 
It's not wrong, though. Mainly confusing.

Reply:
My change on this line only ensures that `late` and `final` are used in the same order as they must be in a variable declaration.

With respect to the confusion caused by the fact that we have two cases:

-  The variable declaration has an initializing expression, so the failing operation is a dynamic invocation of a setter which does not exist (`noSuchMethod`).

- The variable declaration does _not_ have an initializing expression, so the failing operation is an invocation (static or dynamic) of a setter which throws (because there _has_ been a previous write, via a setter invocation or an initializing formal initialization, or an initialization due to a constructor initializer list).

I believe that the current text does cover these cases correctly.

Still, we could reduce the number of logical operations imposed on the reader by making the two relevant cases explicit. I changed the text in order to achieve this.

Note that the new text does not mention the case where there is an initializer: In that case there is no setter, and that is covered by standard rules both in the static and the dynamic case.